File: 0103-binary-tree-zigzag-level-order-traversal/0103-binary-tree-zigzag-level-order-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {

        List<List<Integer>> res=new ArrayList<>();    // Resultant Array bna lo; shuru mei to empty hi hoga

        if(root==null){
            return res;
        }

        Queue<TreeNode> q=new LinkedList<>();
        q.offer(root);

        boolean flag=true;                   // true tab hoga jb direction Left --> Right hogi

        while(!q.isEmpty()){

            int lvlSize=q.size();
            List<Integer> tmp=new ArrayList<>(Collections.nCopies(lvlSize,0));

            int first=0;
            int last=lvlSize-1;

            while(lvlSize-->0){

                TreeNode t=q.poll();

                // Direction ke hisab se store karlo
                if(flag==true){
                    tmp.set(first,t.val);
                    first++;
                }
                else{
                    tmp.set(last,t.val);
                    last--;
                }

                // Left Child ko lo
                if(t.left!=null){
                    q.offer(t.left);
                }

                // Right Child Ko lo
                if(t.right!=null){
                    q.offer(t.right);
                }
            }

            res.add(tmp);

            // Direction Change Karlo
            flag=!flag;
        }

        return res;
    }
}

File: 0103-binary-tree-zigzag-level-order-traversal/0103-binary-tree-zigzag-level-order-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        vector<vector<int>>res;    //Resultant Array bna lo ; shuru mei to empty hi hoga 

        if(root==nullptr){
            return res;
        }
        
        queue<TreeNode*>q;
        q.push(root);

        bool flag=true;                   // true tab hoga jb direction left --> right hogi 

        while(!q.empty()){
            int lvlSize=q.size();
            vector<int>tmp(lvlSize);

            int first=0;
            int last=lvlSize-1;

            while(lvlSize--){
                TreeNode*t=q.front();
                q.pop();

                // Direction ke hisab se store karlo 
                if(flag==true){
                    tmp[first]=t->val;
                    first++;
                }
                else{
                    tmp[last]=t->val;
                    last--;
                }

                // Left Child ko lo 
                if(t->left!=nullptr){
                    q.push(t->left);
                }
                // Right Child Ko lo 
                if(t->right!=nullptr){
                    q.push(t->right);
                }
            }
            res.push_back(tmp);

            // Direction Change Karlo 
            flag=1-flag;
        }
        return res;
    }
};